### PR TITLE
feat(qa_check): enable checking connector docs structure via qa check

### PR DESCRIPTION
**Pull request by darynaishchenko**
## What
as part of https://github.com/airbytehq/airbyte-internal-issues/issues/2887

## How
Added tests for validation critical connectors docs structure to existing `CheckDocumentationStructure` test suite. 
These test were previously implemented in cat https://github.com/airbytehq/airbyte/pull/34380. But cat's structure doesn't allow to pass connector docs file into cat container, so tests were moved to qa checks.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by alafanechere — CHANGES_REQUESTED** [review]
I really appreciate the effort to make our docs more consistent.
But I have some blocking comments:
* We can't reliably get a connector spec before the connector build. The spec source of truth is the `spec` command output, not its json/yaml file. I unfortunately think there's no easy way to implement such a check on the pre-requisite section.
* You hardcoded a "skip the check for some connectors" logic, but you could re-use some existing class attribute we use as flags to only apply the check on a subset of connectors.
* I think the CheckDocumentationStructure has grown too big. It'd benefit maintainability and documentation if you'd declare more granular and new checks. 
* Can you make sure to follow README instruction to introduce a new check. I think you're missing a version bump and the regeneration of the QA checks docs.
* I'd also recommend manually running the `connectors-qa run --connector-directory=airbyte-integrations/connectors` command to make sure the checks are passing for all connectors. To not break connectors' nightly build and parallel efforts I think the failing connectors should be fixed before introducing the check.

**Comment by darynaishchenko** [comment]
@alafanechere, Thanks for the review! 

> I'd also recommend manually running the connectors-qa run --connector-directory=airbyte-integrations/connectors command to make sure the checks are passing for all connectors. To not break connectors' nightly build and parallel efforts I think the failing connectors should be fixed before introducing the check.

Thanks, it was one of my future questions. I'll check it .
I also prepared a pr for fixing connector docs https://github.com/airbytehq/airbyte/pull/39328 I want to merge docs fix before the update for qa checks.

> As I mentioned in an earlier comment, I don't think we have reliable way to consume the connector spec before the connector build or release.

This test suite is now applicable for certified connectors (required fields description is one of certification criteria). I guess we can just skip testing of Prerequisites description if connector doesn't have spec file.

> In other words: the spec source of truth should be the spec command output, not anything declared in spec.yaml / spec.json / manifest

I totally agree with that. But it is a test that is requires both connector image and connector docs file. In cat we can't use docs file and in qa check we can't use spec command(please correct me if we can). For now it's only the facebook-marketing source which requires to run spec command to test Prerequisites description,so this test is being skipped for the source.

**Comment by alafanechere** [comment]
> I guess we can just skip testing of Prerequisites description if connector doesn't have spec file.

@darynaishchenko  That's sounds like a fair workaround to me if you think the file content is reliable enough.
Please add a comment in your code to share this "source of truth" issue. 
If there are too much false positive/negative we can disable this check later.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py]: I don't know such cases. It can be possible only in case when doc file has link to itself, I guess. But workaround for it is adding this links after release in separate pr.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py:337]: divided into `CheckDocumentationHeaders` and `CheckDocumentationDescriptions`

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py:237]: divided into `CheckDocumentationHeaders` and `CheckDocumentationDescriptions`

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py]: added

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/doc_templates/for_airbyte_cloud.txt
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/doc_templates/for_airbyte_cloud.txt:1]: it's possible. txt files are used to strictly compare doc lines with templates and for now we don't need such a big functionality as jinja provides.

**Comment by darynaishchenko** [comment]
@alafanechere I have run qa check on "critical" connectors on my branch with a fix for docs. All docs passed new qa check except google-sheets(1 invalid links was added, and I will fix it in my branch). 
cmd: `airbyte-ci-dev --disable-update-check  connectors --name=source-amazon-ads --name=source-facebook-marketing --name=source-google-ads --name=source-google-analytics-data-api --name=source-google-search-console --name=source-google-sheets --name=source-hubspot --name=source-intercom --name=source-linkedin-ads --name=source-salesforce --name=source-shopify --name=source-stripe --name=source-tiktok-marketing --name=source-zendesk-support test --only-step="qa_checks" `
Also ran `connectors-qa run --connector-directory=airbyte-integrations/connectors`, checks were running as expected but stuck on source-freshservice. It's passed when I run check only for source-freshservice. Do we have an option to skip some connectors when running  connectors-qa run ?

**Review by alafanechere — CHANGES_REQUESTED** [review]
@darynaishchenko can you please fix the units test and type check before I give a final review 🙏

**Comment by darynaishchenko** [comment]
@alafanechere, fixed unit tests and type checks

File: docs/contributing-to-airbyte/resources/qa-checks.md
Comment (alafanechere) [inline docs/contributing-to-airbyte/resources/qa-checks.md:37]: As:
* the check is pretty explicit
* This `hackmd.io` link is not maintained (I think so)
* The `qa-checks.md` is considered as the source of truth for what it takes to have a "clean" connector

**I would highly recommend explicitly stating in the check descripition what is the expected header structure.**

Same applies for `CheckDocumentationDescriptions`.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py:334]: I'd highly recommend using a single Jinja template which we could use to:
* Pre-generate documentation with the correct structure for new connectors 
* Help developers figure what's the expected structure

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py:315]: I'm not convinced that `description` is the right word here. It looks more like "documentation section content", isn't it?

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py:333]: Again, for clarity of checks and documentation I would recommend creating one check per section, you can store the shared logic in `CheckDocumentationContent`

**Review by alafanechere — CHANGES_REQUESTED** [review]
Thanks @darynaishchenko for the changes, it's going in the right direction. Unfortunately I still believe these checks lack explicitness, which is key in the role of this package. I'd love checks to be more granular and better documented so that as a connector developer, if I face a failure, I can easily relate to the qa-checks.md doc to figure how to fix it. And also, as a the maintainer of this package I'd clearly understand the scope and role of this check.

My recommendation would be to ideally:
* Create one check per documentation section you validate
* Use a single template file which stores the expected documentation strucutre, so that we can use it as the source of truth for a valid source structure. It could also be re-used in the connector generator.

I understand it's pretty hard to validate unstructured data like this doc.
The flow could be something like:
* A. Validate the overall doc structure (existence of each section)
* B. If the doc structure is valid we can transform serialize the doc in a more structured data format in memory (a dict with sections as key?)
* C. Feed additional per section checks with the "more structured" documentation.

I believe decoupling parsing and validation would help in boosting the readability of this code.

My other comments are rather not blocking cosmetics changes.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py]: created one check per section, shared logic where it's possible.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation.py]: jinja template was added. also I used it to generate descriptions for some checks, so it now get example from template.md file and puts it to qa-checks doc

File: docs/contributing-to-airbyte/resources/qa-checks.md
Comment (darynaishchenko) [reply docs/contributing-to-airbyte/resources/qa-checks.md:50]: Updated descriptions for check and added examples from template.md as "source of truth"

File: airbyte-ci/connectors/connectors_qa/tests/unit_tests/test_checks/data/docs/invalid_links.md
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/tests/unit_tests/test_checks/data/docs/invalid_links.md:1]: added more invalid links to test doc

File: airbyte-ci/connectors/connectors_qa/tests/unit_tests/test_checks/test_documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/tests/unit_tests/test_checks/test_documentation.py]: added fixtures instead of this method. thanks

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/utils.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/utils.py]: moved to documentation module with helpers and needed models

File: airbyte-ci/connectors/connector_ops/connector_ops/utils.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connector_ops/connector_ops/utils.py]: updated

**Comment by darynaishchenko** [comment]
@alafanechere Thanks for the review! 
> * Create one check per documentation section you validate

Done.

> * Use a single template file which stores the expected documentation strucutre, so that we can use it as the source of truth for a valid source structure. It could also be re-used in the connector generator.

Done.

> * A. Validate the overall doc structure (existence of each section)

It's being validated by `CheckDocumentationHeadersOrder`.

> * B. If the doc structure is valid we can transform serialize the doc in a more structured data format in memory (a dict with sections as key?)

I created Content class which works with both Template Content and Documentation Content. Main methods are 

- headers - all h1-h3 headers from documentation as a list
- sections - dict with header name as a key and list of sections content as value (it's a list because we expect that some headers could have duplicates, for example For Airbyte Cloud). But there was a problem with how to choose section content to validate if we have more than one section so I used `get_close_matches` to find most close content and if it's none check shows expected template content and asks to validate the docs.

Also I added using of `ndiff`  to make error message more user friendly and now it looks like:
image
image

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/__init__.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/__init__.py:3]: 💅 : You could declare `ENABLED_CHECKS` in the `documentation/__init__.py` file to avoid this change.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:130]: I don't get how >=400 status code can be valid...
I think the only valid status code are 200 and redirects.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:154]: Isn't the use of `Thread` overkill here? Why not use `request.get(docs_link, timeout=30)`?

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/templates/template.md
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/templates/template.md:1]: Can you add `.j2` extensions to jinja templates please?

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:221]:
```suggestion
        The order of the headers should follow our standard template
```
The "standard template" is the `template/template.md` now isn't it?

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:205]:
```suggestion
    def get_not_required_headers(self, connector_name: str) -> list[str]:
```

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/templates/documentation_headers_check_description.md
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/templates/documentation_headers_check_description.md:1]: Please do not reference this `hackmd.io`. The work you've done should actually be enough to document the expected structure of documentation right?

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:188]: I'm not sure to get how dynamic this description is. Can we store it inside a static file?

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:303]:
```suggestion
        "The user facing connector documentation should update `Prerequisites`"
        " section with description for all required fields from source specification. "
        "Having described all required fields in a one place helps Airbyte users easily set up the source connector. \n"
        "If spec has required credentials/access_token/refresh_toke etc, "
        'check searches for one of ["account", "auth", "credentials", "access", "client"] words. '
        "No need to add credentials/access_token/refresh_toke etc to the section"
```

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (alafanechere) [inline airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:386]: Again, I don't get why we need a template here. Can't we use a static file describing the expected sections?

**Review by alafanechere — APPROVED** [review]
Thank you so much @darynaishchenko for the effort! 
I've to admit I don't totally get why we're using templates to generate check descriptions.
I've no blocking comment. The code structure is aligned and granular enough to match the other checks, so LGTM.  👍

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:130]: Test performs general unauthorized request, so 401 and 403 response for links that requires auth. I added these statuses because validation of links in some cases requires permission, auth or right method, that we don't know while testing docs. So this statuses can be skipped. Overall we are only interested in 404 not found.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:154]: for some documentations that have a lot of links, for example in changelog section, it makes this test work faster. seconds instead of minutes.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:221]: yes, but it follows https://hackmd.io/Bz75cgATSbm7DjrAqgl4rw. and should be updated if it was changed.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:188]: it only adds section headers and content to description of the check from template. so if template was changed it will be automatically added in description.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py:386]: I just didn't want to duplicate section description here and in template file. For me it's easier when it's generated from template instead of copy pasting as string to description.

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/documentation/documentation.py]: thanks, updated

File: airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/__init__.py
Comment (darynaishchenko) [reply airbyte-ci/connectors/connectors_qa/src/connectors_qa/checks/__init__.py]: updated

**Comment by darynaishchenko** [comment]
@alafanechere, I temporary changed url to standard template in `qa-checks.md`,  Versel check failed because of invalid link: template.md is not found (I think). I'll update it to template.md once these changes are merged.

**Pull request merged**
